google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddAdCustomizer.java

```diff
@@ -0,0 +1,498 @@
+      SearchPagedResponse searchPagedResponse = googleAdsServiceClient.search(request);
+
+      Feed feed = searchPagedResponse.iterateAll().iterator().next().getFeed();
+
+      System.out.printf(
+          "Found the following attributes for feed with name %s:%n",
+          feed.getName().getValue());
+      for(FeedAttribute feedAttribute : feed.getAttributesList()) {
+        System.out.printf(
+            "\t\"%s\" with id %d and type %s%n",
+            feedAttribute.getName().getValue(),
+            feedAttribute.getId().getValue(),
+            feedAttribute.getType());
+        feedAttributes.put(feedAttribute.getName().getValue(), feedAttribute);
+      }
+    }
+    return feedAttributes;
+  }
+
+  /**
+   * Creates a feed mapping to set the feed as an ad customizer feed.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param feedResourceName the resource name of the feed.
+   * @param feedAttributes the attributes of the feed.
+   */
+  private void createAdCustomizerMapping(
+      GoogleAdsClient googleAdsClient, long customerId, String feedResourceName,
+      Map<String, FeedAttribute> feedAttributes) {
+
+    // Map the feed attributes to ad customizer placeholder fields.
+    AttributeFieldMapping nameFieldMapping = AttributeFieldMapping.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Name").getId())
+        .setAdCustomizerField(AdCustomizerPlaceholderField.STRING)
+        .build();
+
+    AttributeFieldMapping priceFieldMapping = AttributeFieldMapping.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Price").getId())
+        .setAdCustomizerField(AdCustomizerPlaceholderField.PRICE)
+        .build();
+
+    AttributeFieldMapping dateFieldMapping = AttributeFieldMapping.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Date").getId())
+        .setAdCustomizerField(AdCustomizerPlaceholderField.DATE)
+        .build();
+
+    FeedMapping feedMapping = FeedMapping.newBuilder()
+        .setFeed(StringValue.of(feedResourceName))
+        // Map the feed to the AD_CUSTOMIZER placeholder type
+        .setPlaceholderType(PlaceholderType.AD_CUSTOMIZER)
+        .addAttributeFieldMappings(nameFieldMapping)
+        .addAttributeFieldMappings(priceFieldMapping)
+        .addAttributeFieldMappings(dateFieldMapping)
+        .build();
+
+    FeedMappingOperation feedMappingOperation = FeedMappingOperation.newBuilder()
+        .setCreate(feedMapping)
+        .build();
+
+    try (FeedMappingServiceClient feedMappingServiceClient =
+        googleAdsClient.getLatestVersion().createFeedMappingServiceClient()) {
+
+      MutateFeedMappingsResponse response = feedMappingServiceClient.mutateFeedMappings(
+          Long.toString(customerId), ImmutableList.of(feedMappingOperation));
+
+      String feedMappingResourceName = response.getResults(0).getResourceName();
```

Comment:
I would just inline this with the logger.

Reply:
Done.